### PR TITLE
[Docs] Cfg job resources

### PR DESCRIPTION
**Pull request by jillnogold**

### CODE REVIEW

File: docs/runtimes/functions.md [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
+addresses scalability and reduces infrastructure costs.
 
-**In this section**
-```{toctree}
-   :maxdepth: 2
+MLRun supports:
+- Real-time functions for: serving, APIs, and stream processing (based on the high-performance Nuclio engine). 
+- Batch functions (based on Kubernetes jobs, Spark, Dask, Horovod, etc.)
+
+Function objects are all inclusive (code, spec, API, and metadata definitions), which allows placing them
```
Comment (yaronha) [inline docs/runtimes/functions.md:14]: need more hyper links in this section, e.g. talk about the fact function resources can be configured and link to the `Managing job resources` page, mention multiple runtimes and link to their doc pages, etc.

File: docs/runtimes/node-affinity.md [hunk 2]
```diff
@@ -1,96 +1,214 @@
-# Node affinity for MLRun jobs
-Node affinity can be applied to MLRun to determine on which nodes 
-they can be placed. The rules are defined using custom labels on nodes and label selectors. 
-Node affinity allows towards Spot or On Demand groups of nodes.
+# Managing job resources
+
+Configuration of job resources is relevant for all supported cloud platforms.
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:3]: Need to start with some overview paragraph, e.g: 

MLRun orchestrate serverless functions over Kubernetes, the user can specify desired resource requirements (CPU, memory, GPUs) and preferences/priorities in the logical function object, this will be used during the function deployment.

File: docs/runtimes/node-affinity.md [hunk 3]
```diff
@@ -1,96 +1,214 @@
...
+  of the function, and select **Edit** and scroll to the **Resources** section.
+-->
 
 **In this section**
-- [On demand vs spot](#On-demand-vs-spot)
-- [Stateless and stateful applications](#Stateless-and-stateful-applications)
-- [Node selector](#node-selector)
+- [Preemption mode: Spot vs. On-demand nodes](#preemption-mode-spot-vs-on-demand-nodes)
+- [Pod priority](#pod-priority)
+- [CPU, GPU, and memory limits for user jobs](#cpu-gpu-and-memory-limits-for-user-jobs)
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:22]: CPU/GPU/.. should be the first section (most widely used), Volumes should be the 2nd, followed by the rest

File: docs/runtimes/functions.md [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
+Function objects are all inclusive (code, spec, API, and metadata definitions), which allows placing them 
+in a shared and versioned function market place. This means that different members of the team can produce or 
+consume functions. Each function is versioned and stored in the MLRun database with a unique hash code, 
+and gets a new hash code upon changes.
+There is also an open [public marketplace](https://www.mlrun.org/marketplace/functions/) that stores many pre-developed functions for
+use in your projects. 
+
+Functions can scale-out across multiple containers. Read more in [**Distributed and Parallel Jobs**](./distributed.md).
 
+<img src="../_static/images/mlrun-functions.png" alt="mlrun-architecture" width="600"/><br>
```
Comment (yaronha) [inline docs/runtimes/functions.md:23]: see updated picture in slack

File: docs/runtimes/node-affinity.md [hunk 5]
```diff
@@ -1,96 +1,214 @@
...
-To assign MLRun jobs to specific nodes you use the Kubernetes node label 
-`app.iguazio.com/lifecycle` with the values of:
+Configure preemption mode by adding the parameter in your Jupyter notebook, specifying a mode from the list of values above. <br>
+For example:
 
-* preemptible – assign to EC2 Spot instances
+```func.with_preemption_mode(mode={value})```
 
-* non-preemptible – assign to EC2 On Demand instances
+See [with_node_selection](api/mlrun.runtimes.#mlrun.runtimes.RemoteRuntime.with_node_selection).
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:72]: why not use proper sphinx method/class links?  also in general need more hyper links  (e.g.: {py:func}`~mlrun.run.get_object`)

File: docs/runtimes/node-affinity.md [hunk 6]
```diff
@@ -1,96 +1,214 @@
...
-command line, and programmable interfaces.
+### SDK configuration
 
-To assign MLRun jobs to specific nodes you use the Kubernetes node label 
-`app.iguazio.com/lifecycle` with the values of:
+Configure preemption mode by adding the parameter in your Jupyter notebook, specifying a mode from the list of values above. <br>
+For example:
 
-* preemptible – assign to EC2 Spot instances
+```func.with_preemption_mode(mode={value})```
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:70]: need more complete example (few lines), at least for the first example (e.g. create a function and set the mode)

File: docs/runtimes/node-affinity.md [hunk 7]
```diff
@@ -1,96 +1,214 @@
...
+Modify the Volumes for an ML function by pressing **ML functions**, then **<img src="../_static/images/kebab-menu.png" width="25"/>** 
+of the function, **Edit** | **Resources** | **Volumes** drop-down list. 
+
+Select the volume mount type: either Auto (using auto-mount), Manual or None. If selecting Manual, fill in the details in the volumes list .for each volume to mount to the pod. Multiple volumes can be configured for a single pod.
+
+### SDK configuration
+
+Configure Volumes attached to a function by using the `apply` function modifier on the function. For example:
+
+```func.apply(mlrun.platforms.mount_v3io())```
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:207]: need more elaborated example, explain the values/options, also for the other cases at least for PVC, and add hyperlinks to the py methods

File: docs/runtimes/node-affinity.md (hunk 5 above)
Comment (jillnogold) [reply docs/runtimes/node-affinity.md]: added

File: docs/runtimes/functions.md (hunk 1 above)
Comment (jillnogold) [reply docs/runtimes/functions.md:14]: added

File: docs/runtimes/functions.md [hunk 8]
```diff
@@ -1,16 +1,41 @@
...
+
+Function objects are all inclusive (code, spec, API, and metadata definitions), which allows placing them 
+in a shared and versioned function market place. This means that different members of the team can produce or 
+consume functions. Each function is versioned and stored in the MLRun database with a unique hash code, 
+and gets a new hash code upon changes.
+
+MLRun supports:
+- [multiple types of runtimes](../concepts/functions-overview).
+- Configuring the function resources (replicas, CPU/GPU/memory limits, volumes, Spot vs. On-demand nodes, pod priority, node affinity). See details in [Managing job resources](node-affinity). 
+- Iterative tasks for automatic and distributed execution of many tasks with variable parameters (hyperparams). See [Hyperparam and iterative jobs](../hyper-params).
```
Comment (yaronha) [inline docs/runtimes/functions.md:22]: hyperparam and distributed runtimes are two separate links

File: docs/runtimes/node-affinity.md [hunk 9]
```diff
@@ -1,96 +1,347 @@
-# Node affinity for MLRun jobs
-Node affinity can be applied to MLRun to determine on which nodes 
-they can be placed. The rules are defined using custom labels on nodes and label selectors. 
-Node affinity allows towards Spot or On Demand groups of nodes.
+# Managing job resources
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:1]: why not also change the `.md` file name?

File: docs/runtimes/node-affinity.md [hunk 10]
```diff
@@ -1,96 +1,347 @@
...
+When creating a job or a function, overwrite the default **Memory**, **CPU**, or **GPU** in the **Configuration** tab, under **Resources**.
+
+### SDK configuration
+
+Configure the limits assigned to a function by using `with_limits`. For example:
+
+```
+import mlrun
+import os
+fn = mlrun.import_function("hub://xgb_trainer")
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:61]: not a good example, this market func will soon be deprecated and too complicated, see cells 8 & 9 in: https://github.com/mlrun/demos/blob/1.0.x/mask-detection/1-training-and-evaluation.ipynb

File: docs/runtimes/node-affinity.md [hunk 11]
```diff
@@ -1,96 +1,347 @@
...
+For example, using v3io storage:
+```
+# import the training function from the marketplace (hub://)
+train = mlrun.import_function('hub://sklearn_classifier')# Import the function:
+open_archive_function = mlrun.import_function("hub://open_archive")
+
+# use mount_v3io() for iguazio volumes
+open_archive_function.apply(mount_v3io())
+```
+You can specify the v3io path to use for the volume (`remote`), the volume mount path (`volume_mounts`), and security credentials, for example: `mlrun.platforms.mount_v3io(name='v3io', remote='', mount_path='', access_key='', user='', secret=None, volume_mounts=None)`. See full details in [mount_v3io](../api/mlrun.platforms.html#mlrun.platforms.mount_v3io).
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:123]: You can specify a list of the v3io path to use and how they map inside the container ( using `volume_mounts`)... for example: `mlrun.mount_v3io(name='data', access_key='XYZ123..', volume_mounts=[mlrun.VolumeMount("/data", "projects/proj1/data")])`

File: docs/runtimes/node-affinity.md [hunk 12]
```diff
@@ -1,96 +1,347 @@
...
+
+Alternatively, using a PVC volume:
+```
+# use or mount_pvc() for k8s PVC volumes
+open_archive_function.apply(mount_pvc())
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:128]: mount_pvc(pvc_name="data-claim", volume_name="data", volume_mount_path="/data")

File: docs/runtimes/node-affinity.md [hunk 13]
```diff
@@ -1,96 +1,347 @@
...
+
+### SDK configuration
+
+Configure preemption mode by adding the parameter in your Jupyter notebook, specifying a mode from the list of values above. <br>
+This example illustrates a function that cannot be scheduled on preemptible nodes:
+
+```
+import mlrun
+import os
+fn = mlrun.import_function("hub://xgb_trainer")
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:180]: not a good example, old and too long, use this instead:

```
train_fn = mlrun.code_to_function('training', 
                            kind='job',
                            handler='my_training_function')
train_fn.with_preemption_mode(mode="prevent")
train_fn.run(inputs={"dataset"  :my_data})

File: docs/runtimes/node-affinity.md [hunk 14]
```diff
@@ -1,96 +1,347 @@
...
+Modify the priority for an ML function by pressing **ML functions**, then **<img src="../_static/images/kebab-menu.png" width="25"/>** 
+of the function, **Edit** | **Resources** | **Pods Priority** drop-down list.
+
+
+### SDK configuration
+
+Configure pod priority by adding the priority class parameter in your Jupyter notebook. <br>
+For example:
+
+```
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:261]: we dont need the full example or use a short one like above

File: docs/runtimes/node-affinity.md [hunk 15]
```diff
@@ -1,96 +1,347 @@
...
+selector**, and adding or removing Key:Value pairs. Configure the node selection for individual Nuclio functions when creating a function in 
+the **Confguration** tab, under **Resources**, by adding Key:Value pairs.
+
+### SDK configuration
+
+Configure node selection by adding the key:value pairs in your Jupyter notebook formatted as a Python dictionary. <br>
+For example:
+
+```
+import mlrun
```
Comment (yaronha) [inline docs/runtimes/node-affinity.md:323]: same

File: docs/runtimes/node-affinity.md (hunk 10 above)
Comment (jillnogold) [reply docs/runtimes/node-affinity.md:61]: This section should show example of GPU, limits, and memory. The one you pointed to is only GPUs. How to add memory and limits to this example?
# If you wantto train on gpu, set this variable to 'True', otherwise 'False':
use_gpu = False

# Create the MLRun Function:
mpi_training_and_evaluation_function = mlrun.code_to_function(
    filename=os.path.join(framework, "training-and-evaluation.py"),
    name="mpi-training-and-evaluation",
    handler="train",
    kind="mpijob",
    image="mlrun/ml-models-gpu" if use_gpu else "mlrun/ml-models",
    with_doc=False
)

# Setup the desired configurations:
mpi_training_and_evaluation_function.spec.replicas = 2
if use_gpu:
    # Select the number of GPUs per replica:
    mpi_training_and_evaluation_function.gpus(1)
else:
    mpi_training_and_evaluation_function.with_requests(cpu=2)

Comment (yaronha) [reply docs/runtimes/node-affinity.md]: @jillnogold keep it simple:

```python
training_function = mlrun.code_to_function("training.py", name="training", handler="train", 
                                                                       kind="mpijob", image="mlrun/ml-models-gpu")
training_function.spec.replicas = 2
training_function.with_requests(cpu=2)
training_function.gpus(1)
```

File: docs/runtimes/functions.md (hunk 8 above)
Comment (jillnogold) [reply docs/runtimes/functions.md:22]: the xref in line 22 also includes Parallel execution over containers. Do you want another xref to the topic?

**Comment by yaronha** [comment]
Approved, LGTM

File: docs/runtimes/cfg-job-resources.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+  You can configure how much of each resource a function needs. Kubernetes uses this information when placing a pod on a node. The memory and 
+  CPU configurations that you specify in the service are applied to each replica. Limits and requests are supported for all services.
+
+  Service limit!
+
+  When creating a new function, set the **Memory** and **CPU** in the **Common Parameters** tab, under **Resources** .
+
+  Modify the Memory, CPU, GPU for an ML function by pressing **ML functions**, then press **<img src="../_static/images/kebab-menu.png" width="25"/>** 
+  of the function, and select **Edit** and scroll to the **Resources** section.
+-->
```
Comment (Hedingber) [inline docs/runtimes/cfg-job-resources.md:20]: Remove ?

File: docs/runtimes/cfg-job-resources.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+spec.max_replicas = 4
+```
+See more details in [Dask](../runtimes/dask-overview), [MPIJob and Horovod](../runtimes/horovod), [Spark](../runtimes/spark-operator), [Nuclio](../concepts/nuclio-real-time-functions).
+
+## CPU, GPU, and memory limits for user jobs  
+
+When you create a pod in an MLRun job or Nuclio function, the pod has default CPU and memory limits. When the job runs, it can consume 
+resources up to the limits defined. The default limits are set at the service level. You can change the default limit for the service, and 
+also overwrite the default when creating a job, or a function. 
+<!-- The default values are: -->
```
Comment (Hedingber) [inline docs/runtimes/cfg-job-resources.md:46]: same

**Review by Hedingber — CHANGES_REQUESTED** [review]
